tests/sentry/utils/test_pipeline.py

```diff
@@ -0,0 +1,95 @@
+from __future__ import absolute_import
+
+from django.http import HttpRequest
+
+from sentry.testutils import TestCase
+from sentry.utils.pipeline import (
+    PipelineableProvider,
+    PipelineView,
+    ProviderPipeline,
+)
+
+
+class PipelineStep(PipelineView):
+    def dispatch(self, request, pipeline):
+        pipeline.dispatch_count += 1
+        pipeline.bind_state('some_state', 'value')
+
+
+class DummyProvider(PipelineableProvider):
+    key = 'dummy'
+    pipeline = [PipelineStep(), PipelineStep()]
+
+    def get_pipeline(self):
+        return self.pipeline
+
+
+class DummyProviderManager(object):
+    def get(self, provider):
+        return DummyProvider()
+
+
+class DummpyPipeline(ProviderPipeline):
+    pipeline_name = 'test_pipeline'
+    provider_manager = DummyProviderManager()
+
+    def finish_pipeline(self):
+        self.finished = 1
+
+
+class PipelineTestCase(TestCase):
+    def test_simple_pipeline(self):
+        org = self.create_organization()
+        request = HttpRequest()
+        request.session = {}
+        request.user = self.user
+
+        pipeline = DummpyPipeline(request, org, 'dummy', {'some_config': True})
+        pipeline.initialize()
+
+        assert pipeline.is_valid()
+
+        assert 'some_config' in pipeline.provider.config
+
+        pipeline.finished = 0
+        pipeline.dispatch_count = 0
+
+        # Piepline has two steps, ensure both steps compete.
+        # Usually the dispatch itself would be the one calling the current_step
+        # and next_step methods after it determins if it can move forward a
+        # step.
+        pipeline.current_step()
+        assert pipeline.dispatch_count == 1
+        assert pipeline.fetch_state('some_state') == 'value'
+
+        pipeline.next_step()
+        assert pipeline.dispatch_count == 2
+
+        pipeline.next_step()
```

Comment:
It seems a little strange to have to call `current_step` once and then `next_step` subsequently for this pipeline. It might be nicer if the logic in those functions was set up so that you could just do:
```
while pipeline.next_step():
    ...
```

Reply:
ah I see your comment about dispatch

Reply:
This test is a little bit convoluted from usually usage unfortunately. Usually the pipeline view itself would be calling `next_step` and then the initiating view that gets / creates the pipeline would call `current_step`.